psl-core/src/main/java/org/linqs/psl/application/inference/online/OnlineInference.java

```diff
@@ -0,0 +1,309 @@
+        }
+    }
+
+    protected void doDeleteAtom(DeleteAtom nextAction) throws IllegalArgumentException {
+        // Resolve Predicate
+        Predicate registeredPredicate = Predicate.get(nextAction.getPredicateName());
+        if (registeredPredicate == null) {
+            throw new IllegalArgumentException("Predicate is not registered: " + nextAction.getPredicateName());
+        }
+
+        ((OnlineTermStore)termStore).deleteAtom(registeredPredicate, nextAction.getArguments());
+    }
+
+    protected void doUpdateObservation(UpdateObservation nextAction) throws IllegalArgumentException {
+        // Resolve Predicate
+        Predicate registeredPredicate = Predicate.get(nextAction.getPredicateName());
+        if (registeredPredicate == null) {
+            throw new IllegalArgumentException("Predicate is not registered: " + nextAction.getPredicateName());
+        }
+
+        ((OnlineTermStore)termStore).updateAtom(registeredPredicate, nextAction.getArguments(), nextAction.getValue());
+    }
+
+    protected void doClose(Close nextAction) {
+        close = true;
+    }
+
+    protected void doWriteInferredPredicates(WriteInferredPredicates nextAction) {
+        long start;
+        long stop;
+        // Activate the atoms that were added
+        log.trace("Partial Grounding Start");
+        start = System.currentTimeMillis();
+        ArrayList<GroundRule> groundRules = ((OnlineAtomManager)atomManager).activateAtoms(rules, (OnlineTermStore) termStore);
+        stop = System.currentTimeMillis();
+        log.trace("Partial Grounding Stop");
+        log.trace("Partial Grounding Time: {}", (stop - start));
+
+        log.trace("Term Generation Start");
+        log.trace("Adding " + groundRules.size() + " ground rules to model");
+        start = System.currentTimeMillis();
+        int newTermCount = 0;
+        for (GroundRule groundRule : groundRules) {
+            ReasonerTerm newTerm = termGenerator.createTerm(groundRule, termStore);
+            if (newTerm == null) {
+                // Term was trivial
+                continue;
+            }
+            newTermCount++;
+            termStore.add(groundRule, newTerm);
+        }
+        stop = System.currentTimeMillis();
+        log.trace("Added " + newTermCount + " terms to model");
+        log.trace("Term Generation Stop");
+        log.trace("Term Generation Time: {}", (stop - start));
+
+        // Ensure we are in optimal state
+        log.trace("Optimization Start");
+        start = System.currentTimeMillis();
+        objective = reasoner.optimize(termStore);
+        stop = System.currentTimeMillis();
+        log.trace("Optimization Start");
+        log.trace("Optimization Time: {}", (stop - start));
+
+        outputResults(nextAction.getOutputDirectoryPath());
+    }
+
+    private void outputResults(String outputDirectoryPath) {
+        // TODO: (Charles) Duplicated code fragment from launcher.
```

Comment:
You can make a common method in Database that outputs all the random variables to an output stream.
Then the CLI can call that as well.